src/crypto/sha256.cpp

```diff
@@ -716,6 +716,63 @@ void CSHA256::Midstate(unsigned char hash[OUTPUT_SIZE], uint64_t* len, unsigned
     }
 }
 
+std::vector<unsigned char> CSHA256::Save() {
+   std::vector<unsigned char> result;
+   result.reserve(40 + (bytes % 64));
+
+   for(auto i = 0; i < 8; ++i) {
+     result.push_back(static_cast<unsigned char>(s[i] >> 24));
+     result.push_back(static_cast<unsigned char>(s[i] >> 16));
+     result.push_back(static_cast<unsigned char>(s[i] >>  8));
+     result.push_back(static_cast<unsigned char>(s[i]      ));
+   }
+
+   uint64_t bits = static_cast<uint64_t>(bytes) << 3;
+   result.push_back(static_cast<unsigned char>(bits      ));
+   result.push_back(static_cast<unsigned char>(bits >>  8));
+   result.push_back(static_cast<unsigned char>(bits >> 16));
+   result.push_back(static_cast<unsigned char>(bits >> 24));
+   result.push_back(static_cast<unsigned char>(bits >> 32));
+   result.push_back(static_cast<unsigned char>(bits >> 40));
+   result.push_back(static_cast<unsigned char>(bits >> 48));
+   result.push_back(static_cast<unsigned char>(bits >> 56));
+
+   size_t buf_size = bytes % 64;
+   for(size_t i = 0; i < buf_size; ++i) {
+      result.push_back(buf[i]);
+   }
+
+   return result;
+}
+
+bool CSHA256::Load(const std::vector<unsigned char>& vch) {
+    if (vch.size() < 40) return false;
+
+    uint64_t bits = static_cast<uint64_t>(vch[32])
+                  | static_cast<uint64_t>(vch[33]) << 8
+                  | static_cast<uint64_t>(vch[34]) << 16
+                  | static_cast<uint64_t>(vch[35]) << 24
+                  | static_cast<uint64_t>(vch[36]) << 32
+                  | static_cast<uint64_t>(vch[37]) << 40
+                  | static_cast<uint64_t>(vch[38]) << 48
+                  | static_cast<uint64_t>(vch[39]) << 56;
+
+    if ((bits & 0x07) != 0 || vch.size() != 40 + ((bits >> 3) % 64)) return false;
+
+    bytes = bits >> 3;
```

Comment:
nitpick: you can calculate `bytes` before the previous statement and then use `bytes` instead of `bits >> 3` in the size check

Reply:
I'm torn on the issue, because it means that the 3 most significant bits of the `uint64_t` end up getting truncated and end up having no semantic value.  I don't know how I feel about that.  Depending on how people use these new opcodes, it could potentiality become a source of witness malleability.

Honestly I see myself being swayed to using bytes eventually, but I'm going to resist just a little while longer.

Reply:
the 3 least significant bits are checked with `bits & 0x07`, they are not ignored. What I had in mind is
```
bytes = bits >> 3;

if ((bits & 0x07) != 0 || vch.size() != 40 + (bytes % 64)) return false;
```

Reply:
Granted, it is slightly less readable because of bytes/bits visual similarity.

Reply:
might be more readable as two separate checks

```
if ((bits & 0x07) != 0) return false;

bytes = bits >> 3;

if (vch.size() != 40 + (bytes % 64)) return false;
```

Reply:
Ah.  I deliberately wanted to leave the internal state of the CSHA256 unchanged if the `false` is returned here.  Clearly this is worth adding a comment to the code.

Reply:
definitely.